subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/AbstractIncompleteSnapshotWithChildrenTest.groovy

```diff
@@ -0,0 +1,417 @@
+        childPaths.collect {
+            new VirtualFileSystemTestSpec(childPaths, it, 0, it)
+        }
+    }
+    static final List<VirtualFileSystemTestSpec> DESCENDANT_PATH = CHILD_CONSTELLATIONS.collectMany { childPaths ->
+        childPaths.collect {
+            new VirtualFileSystemTestSpec(childPaths, "${it}/descendant", 0, it)
+        }
+    }
+
+    def getSelectedChildSnapshot(VirtualFileSystemTestFixture fixture, @Nullable foundSnapshot = fixture.selectedChild) {
+        1 * fixture.selectedChild.getSnapshot(fixture.absolutePath, fixture.absolutePath.length() + 1) >> Optional.ofNullable(foundSnapshot)
+    }
+
+    def getDescendantSnapshotOfSelectedChild(VirtualFileSystemTestFixture fixture, @Nullable MetadataSnapshot foundSnapshot) {
+        def descendantOffset = fixture.offset + fixture.selectedChild.pathToParent.length() + 1
+        1 * fixture.selectedChild.getSnapshot(fixture.absolutePath, descendantOffset) >> Optional.ofNullable(foundSnapshot)
+    }
+
+    def storeDescendantOfSelectedChild(VirtualFileSystemTestFixture fixture, MetadataSnapshot snapshot, FileSystemNode updatedChild) {
+        def descendantOffset = fixture.offset + fixture.selectedChild.pathToParent.length() + 1
+        1 * fixture.selectedChild.store(fixture.absolutePath, descendantOffset, snapshot) >> updatedChild
+    }
+
+    def invalidateDescendantOfSelectedChild(VirtualFileSystemTestFixture fixture, @Nullable FileSystemNode invalidatedChild) {
+        def descendantOffset = fixture.offset + fixture.selectedChild.pathToParent.length() + 1
+        1 * fixture.selectedChild.invalidate(fixture.absolutePath, descendantOffset) >> Optional.ofNullable(invalidatedChild)
+    }
+
+    @SuppressWarnings("GrMethodMayBeStatic")
+    void noMoreInteractions() {
+        _ * _.pathToParent
+        0 * _
+    }
+
+    private static String findPathWithParent(List<String> childPaths, String parentPath) {
+        childPaths.find { PathUtil.isChildOfOrThis(it, 0, parentPath) }
+    }
+
+    private static List<String> parentPaths(String childPath) {
+        (childPath.split('/') as List).inits().tail().init().collect { it.join('/') }
+    }
+
+
+    static List<FileSystemNode> sortedChildren(FileSystemNode... children) {
+        def childrenOfIntermediate = children as List
+        childrenOfIntermediate.sort(Comparator.comparing({ FileSystemNode node -> node.pathToParent }, PathUtil.pathComparator()))
+        childrenOfIntermediate
+    }
+
+    FileSystemNode mockChild(String pathToParent) {
+        Mock(FileSystemNode, defaultResponse: new RespondWithPathToParent(pathToParent), name: pathToParent)
+    }
+
+    def <T extends FileSystemNode> T mockSnapshot(
+        Class<T> type,
+        String pathToParent
+    ) {
+        Mock(type, defaultResponse: new RespondWithPathToParent(pathToParent), name: pathToParent)
+    }
+
+    MetadataSnapshot mockSnapshot(String pathToParent) {
+        mockSnapshot(MetadataSnapshot, pathToParent)
+    }
```

Comment:
These mocks (especially taking `RespondWithPathToParent` into account) look more complicated than they should necessarily be. Can we simplify them somehow?